tutorial/lesson_18_parallel_reductions.cpp

```diff
@@ -0,0 +1,219 @@
+        // condition. Note, however, the histogram operation (i.e. the sum reduction)
+        // is associative. A common trick to speed-up this type of reduction is to
+        // split the reduction into chunks. Within a chunk, the reduction is not
+        // serial, however, since chunks are independent of each other, we can
+        // parallelize in chunk granularity.
+        //
+        // Going back to the histogram example, we split the reduction into chunks
+        // by defining an intermediate function:
+        //
+        // Func intm("intm");
+        // Var i("i"), y("y");
+        // intm(i, y) = 0;
+        // RDom rx(0, input.width());
+        // intm(input(rx, y), y) += 1;
+        //
+        // This intermediate computes the sum reductions across the x dimension
+        // (denoted by rx) for each y dimension independently. The histogram function
+        // now sums over the partial results across the x dimension (denoted by i)
+        // computed by the intermediate function:
+        //
+        // Func histogram("histogram");
+        // histogram(i) = 0;
+        // RDom ry(0, input.height());
+        // histogram(i) += intm(i, ry);
+        //
+        // Since the intermediate no longer has data dependencies across the y dimension,
+        // we can parallelize it over y:
+        // intm.compute_root().update().parallel(y);
+        //
+        // This manual factorization of an associative reduction can be tedious and
+        // bug-prone. Although it's fairly easy to do manually for histogram,
+        // it can get complex pretty fast -- RDom may have predicates defined over
+        // it or the value computed may be a multi-dimensional tuple. Halide
+        // provides a way to do this type of factorization through the scheduling
+        // directive 'rfactor'. rfactor splits an associative update definition
+        // into an intermediate which computes the partial results over slices of a
+        // reduction domain and replace the current update definition with a new
+        // definition which merges those partial results.
+
+        // Using rfactor, we don't need to change the algorithm at all:
+        Func histogram("histogram");
+        histogram(x) = 0;
+        RDom r(0, input.width(), 0, input.height());
+        histogram(input(r.x, r.y)) += 1;
+
+        // The factoring of associative reduction is moved into the schedule,
+        // via rfactor. To generate the same code as the manually-factored
+        // version, we do the following:
+        Func intm = histogram.update().rfactor({{r.y, y}});
+        // rfactor takes as input a list of <RVar, Var> pairs, which contains all
+        // reduction variables to be removed from the original function and lifted
+        // to the intermediate function. The remaining reduction variables are
+        // made 'pure' in the intermediate function, which make them race-condition
+        // free and hence are parallelizable.
```

Comment:
I think it would be good to spell out what the RVar and Var are used for a bit more here.